File: cmd/ingest.go
package cmd

import (
	"encoding/json"
	"fmt"
	"github.com/samahlstrom/forge-cli/internal/ui"
	"github.com/samahlstrom/forge-cli/internal/util"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/spf13/cobra"
)

var (
	ingestChunkSize string
	ingestResume    string
)

// SpecAnalysis holds the structured output from Claude's spec analysis.
type SpecAnalysis struct {
	ProjectName    string   `json:"project_name"`
	Description    string   `json:"description"`
	Language       string   `json:"language"`
	Framework      *string  `json:"framework"`
	ProjectType    string   `json:"project_type"`
	Modules        []string `json:"modules"`
	Architecture   string   `json:"architecture"`
	SensitiveAreas string   `json:"sensitive_areas"`
	DomainRules    string   `json:"domain_rules"`
	Constraints    []string `json:"constraints"`
	PageCount      *int     `json:"page_count"`
}

type resolvedFile struct {
	path        string
	ext         string
	name        string
	sizeDisplay string
	sizeBytes   int64
	pageCount   *int
}

func init() {
	ingestCmd := &cobra.Command{
		Use:   "ingest <files...>",
		Short: "Ingest one or more spec documents for project planning and decomposition",
		Args:  cobra.MinimumNArgs(1),
		RunE:  runIngest,
	}
	ingestCmd.Flags().StringVar(&ingestChunkSize, "chunk-size", "20", "Pages per chunk for PDF processing")
	ingestCmd.Flags().StringVar(&ingestResume, "resume", "", "Resume analysis of an existing spec")
	rootCmd.AddCommand(ingestCmd)
}

func runIngest(cmd *cobra.Command, args []string) error {
	cwd, _ := os.Getwd()

	ui.Intro(ui.Bold("forge") + ui.Dim(" — Spec Ingestion"))

	// ── Resolve and validate all spec files ──
	supportedFormats := map[string]bool{".pdf": true, ".md": true, ".txt": true, ".markdown": true}
	var resolved []resolvedFile

	for _, file := range args {
		specPath, err := filepath.Abs(file)
		if err != nil {
			ui.Cancel(fmt.Sprintf("Invalid path: %s", file))
			os.Exit(1)
		}
		if !util.Exists(specPath) {
			ui.Cancel(fmt.Sprintf("File not found: %s", specPath))
			os.Exit(1)
		}

		ext := strings.ToLower(filepath.Ext(specPath))
		if !supportedFormats[ext] {
			supported := []string{".pdf", ".md", ".txt", ".markdown"}
			ui.Cancel(fmt.Sprintf("Unsupported format: %s (%s). Supported: %s",
				ext, filepath.Base(specPath), strings.Join(supported, ", ")))
			os.Exit(1)
		}

		info, err := os.Stat(specPath)
		if err != nil {
			ui.Cancel(fmt.Sprintf("Cannot stat file: %s", specPath))
			os.Exit(1)
		}
		sizeBytes := info.Size()
		var sizeDisplay string
		if sizeBytes < 102400 {
			sizeDisplay = fmt.Sprintf("%d KB", sizeBytes/1024)
		} else {
			sizeDisplay = fmt.Sprintf("%.1f MB", float64(sizeBytes)/(1024*1024))
		}

		var pageCount *int
		if ext == ".pdf" {
			pageCount = detectPageCount(specPath)
		}

		resolved = append(resolved, resolvedFile{
			path:        specPath,
			ext:         ext,
			name:        filepath.Base(specPath),
			sizeDisplay: sizeDisplay,
			sizeBytes:   sizeBytes,
			pageCount:   pageCount,
		})
	}

	isMulti := len(resolved) > 1

	// ── Display source info ──
	var infoLines []string
	if isMulti {
		infoLines = append(infoLines, fmt.Sprintf("Documents: %s", ui.Cyan(fmt.Sprint(len(resolved)))))
		for _, f := range resolved {
			format := extToFormat(f.ext)
			pageInfo := ""
			if f.pageCount != nil {
				pageInfo = fmt.Sprintf(" (%dp)", *f.pageCount)
			}
			infoLines = append(infoLines, fmt.Sprintf("  %s %s %s",
				ui.Dim("\u2022"), ui.Cyan(f.name), ui.Dim(fmt.Sprintf("%s, %s%s", format, f.sizeDisplay, pageInfo))))
		}
	} else {
		f := resolved[0]
		format := extToFormatLong(f.ext)
		infoLines = append(infoLines, fmt.Sprintf("File:     %s", ui.Cyan(f.name)))
		infoLines = append(infoLines, fmt.Sprintf("Format:   %s", ui.Cyan(format)))
		infoLines = append(infoLines, fmt.Sprintf("Size:     %s", ui.Cyan(f.sizeDisplay)))
		if f.pageCount != nil {
			infoLines = append(infoLines, fmt.Sprintf("Pages:    %s", ui.Cyan(fmt.Sprint(*f.pageCount))))
		}
	}

	chunkSize, _ := strconv.Atoi(ingestChunkSize)
	if chunkSize <= 0 {
		chunkSize = 20
	}
	totalPages := 0
	for _, f := range resolved {
		if f.pageCount != nil {
			totalPages += *f.pageCount
		}
	}
	if totalPages > chunkSize {
		chunks := (totalPages + chunkSize - 1) / chunkSize
		infoLines = append(infoLines, fmt.Sprintf("Chunks:   %s", ui.Cyan(fmt.Sprintf("%d x %d pages", chunks, chunkSize))))
	}

	noteTitle := "Source Document"
	if isMulti {
		noteTitle = "Source Documents"
	}
	ui.Note(strings.Join(infoLines, "\n"), noteTitle)

	// ── Generate spec ID and create directory ──
	specID := "spec-" + util.RandomHex(4)
	specDir := filepath.Join(cwd, ".forge", "specs", specID)
	util.EnsureDir(specDir)

	sourceFiles := make([]string, len(resolved))
	for i, f := range resolved {
		var destName string
		if isMulti {
			destName = fmt.Sprintf("source-%d%s", i+1, f.ext)
		} else {
			destName = "source" + f.ext
		}
		util.CopyFile(f.path, filepath.Join(specDir, destName))
		sourceFiles[i] = destName
	}

	// ── If multiple text/markdown files, create combined.md ──
	var combinedPath string
	if isMulti {
		var parts []string
		for _, f := range resolved {
			if f.ext == ".pdf" {
				parts = append(parts, fmt.Sprintf("\n\n---\n# [PDF Document: %s]\n# (Read separately via PDF reader)\n---\n\n", f.name))
			} else {
				content, err := util.ReadText(f.path)
				if err != nil {
					content = ""
				}
				parts = append(parts, fmt.Sprintf("\n\n---\n# Document: %s\n---\n\n%s", f.name, content))
			}
		}
		combinedPath = filepath.Join(specDir, "combined.md")
		_ = util.WriteText(combinedPath, strings.Join(parts, ""))
	}

	suffix := ""
	if isMulti {
		suffix = "s"
	}
	ui.Log.Success(fmt.Sprintf("Copied %d file%s to %s", len(resolved), suffix, ui.Dim(fmt.Sprintf(".forge/specs/%s/", specID))))
	if combinedPath != "" {
		ui.Log.Success(fmt.Sprintf("Combined document created at %s", ui.Dim(fmt.Sprintf(".forge/specs/%s/combined.md", specID))))
	}

	// ── Write spec metadata ──
	type metaSource struct {
		Original string   `json:"original"`
		Stored   string   `json:"stored"`
		Format   string   `json:"format"`
		SizeMB   float64  `json:"size_mb"`
		Pages    *int     `json:"pages"`
	}
	type metaDoc struct {
		SpecID string `json:"spec_id"`
		Source struct {
			Files      []metaSource `json:"files"`
			Combined   *string      `json:"combined"`
			IngestedAt string       `json:"ingested_at"`
		} `json:"source"`
		Status    string `json:"status"`
		ChunkSize int    `json:"chunk_size"`
	}

	meta := metaDoc{
		SpecID:    specID,
		Status:    "pending-analysis",
		ChunkSize: chunkSize,
	}
	meta.Source.IngestedAt = time.Now().UTC().Format(time.RFC3339)
	if combinedPath != "" {
		c := "combined.md"
		meta.Source.Combined = &c
	}
	for i, f := range resolved {
		meta.Source.Files = append(meta.Source.Files, metaSource{
			Original: f.name,
			Stored:   sourceFiles[i],
			Format:   strings.TrimPrefix(f.ext, "."),
			SizeMB:   float64(int(float64(f.sizeBytes)/(1024*1024)*100)) / 100,
			Pages:    f.pageCount,
		})
	}
	metaJSON, _ := json.MarshalIndent(meta, "", "  ")
	_ = util.WriteText(filepath.Join(specDir, "meta.json"), string(metaJSON))

	// ── Spec analysis (if no harness exists) ──
	harnessExists := util.Exists(filepath.Join(cwd, "forge.yaml"))
	var analysis *SpecAnalysis

	if !harnessExists {
		spinner := ui.NewSpinner()
		spinner.Start("Analyzing spec with Claude Code...")

		analysisTarget := filepath.Join(specDir, sourceFiles[0])
		analysisExt := resolved[0].ext
		analysisPagesCount := resolved[0].pageCount
		if combinedPath != "" {
			analysisTarget = combinedPath
			analysisExt = ".md"
			analysisPagesCount = nil
		}

		a, err := analyzeSpecWithClaude(analysisTarget, analysisExt, analysisPagesCount, "")
		if err != nil {
			spinner.Stop("Spec analysis failed")
			ui.Log.Warn("Could not analyze spec automatically. You can configure manually.")
			ui.Log.Warn(ui.Dim(err.Error()))
		} else {
			spinner.Stop("Spec analysis complete")
			analysis = a
		}

		if analysis != nil {
			extractedLines := []string{
				fmt.Sprintf("Project:      %s", analysis.ProjectName),
				fmt.Sprintf("Description:  %s", analysis.Description),
				fmt.Sprintf("Type:         %s", analysis.ProjectType),
				fmt.Sprintf("Language:     %s", analysis.Language),
			}
			if analysis.Framework != nil && *analysis.Framework != "" {
				extractedLines = append(extractedLines, fmt.Sprintf("Framework:    %s", *analysis.Framework))
			}
			if len(analysis.Modules) > 0 {
				extractedLines = append(extractedLines, fmt.Sprintf("Modules:      %s", strings.Join(analysis.Modules, ", ")))
			}
			extractedLines = append(extractedLines, fmt.Sprintf("Architecture: %s", analysis.Architecture))
			if analysis.SensitiveAreas != "" {
				extractedLines = append(extractedLines, fmt.Sprintf("Sensitive:    %s", analysis.SensitiveAreas))
			}
			if len(analysis.Constraints) > 0 {
				max := 3
				if len(analysis.Constraints) < max {
					max = len(analysis.Constraints)
				}
				extractedLines = append(extractedLines, fmt.Sprintf("Constraints:  %s", strings.Join(analysis.Constraints[:max], "; ")))
			}

			ui.Note(strings.Join(extractedLines, "\n"), "Extracted from spec")

			confirmed, cancelled := ui.Confirm("Does this look right?", true)
			if cancelled {
				ui.Cancel("Cancelled.")
				os.Exit(0)
			}

			if !confirmed {
				corrections, cancelled := ui.Text("What needs to change?", "e.g. Use SvelteKit instead of Next.js, add billing as a module")
				if cancelled {
					ui.Cancel("Cancelled.")
					os.Exit(0)
				}
				if corrections != "" {
					spinner2 := ui.NewSpinner()
					spinner2.Start("Re-analyzing with corrections...")
					a2, err := analyzeSpecWithClaude(analysisTarget, analysisExt, analysisPagesCount, corrections)
					if err != nil {
						spinner2.Stop("Re-analysis failed, using original analysis")
					} else {
						spinner2.Stop("Updated analysis complete")
						analysis = a2
					}
				}
			}

			analysisJSON, _ := json.MarshalIndent(analysis, "", "  ")
			_ = util.WriteText(filepath.Join(specDir, "analysis.json"), string(analysisJSON))
			ui.Log.Success("Spec analysis saved.")
		}
	}

	// ── Offer to run init ──
	if !harnessExists {
		runInit, cancelled := ui.Confirm("Scaffold the harness now?", true)
		if cancelled {
			ui.Cancel("Cancelled.")
			os.Exit(0)
		}

		if runInit {
			ui.Outro(ui.Green("Spec ingested — starting init..."))
			return RunInit(analysis, specID)
		}
	}

	// ── Next steps ──
	if harnessExists {
		ui.Log.Step("Next step:")
		ui.Log.Message(fmt.Sprintf("  Open Claude Code -> %s", ui.Cyan(fmt.Sprintf("/ingest %s", specID))))
	} else {
		ui.Log.Step("Next steps:")
		ui.Log.Message(fmt.Sprintf("  1. Run %s to scaffold the harness", ui.Cyan("forge init")))
		ui.Log.Message(fmt.Sprintf("     %s", ui.Dim("(use the spec analysis above to answer onboarding questions)")))
		ui.Log.Message(fmt.Sprintf("  2. Open Claude Code -> %s", ui.Cyan(fmt.Sprintf("/ingest %s", specID))))
	}

	ui.Outro(ui.Green("Spec ready for analysis."))
	return nil
}

// RunInit is called from ingest to launch the init flow with pre-filled spec analysis.
// This is a placeholder that should be wired to the real init command once cmd/init.go exists.
func RunInit(analysis *SpecAnalysis, specID string) error {
	// Build args for the init command invocation
	initArgs := []string{"--spec-id", specID}
	if analysis != nil {
		// Write analysis to a temp file so init can read it
		tmpFile := filepath.Join(os.TempDir(), fmt.Sprintf("forge-analysis-%s.json", util.RandomHex(4)))
		data, _ := json.MarshalIndent(analysis, "", "  ")
		_ = util.WriteText(tmpFile, string(data))
		defer os.Remove(tmpFile)
		initArgs = append(initArgs, "--spec-analysis", tmpFile)
	}

	// Find and execute the init subcommand on rootCmd
	initCmd, _, err := rootCmd.Find([]string{"init"})
	if err != nil || initCmd == rootCmd {
		// init command not yet registered; fall back to exec
		ui.Log.Warn("Init command not available. Run `forge init` manually.")
		return nil
	}
	initCmd.SetArgs(initArgs)
	return initCmd.Execute()
}

// ─── Spec analysis via Claude subprocess ──────────────────────────

func analyzeSpecWithClaude(specPath, ext string, pageCount *int, corrections string) (*SpecAnalysis, error) {
	prompt := buildAnalysisPrompt(specPath, ext, pageCount, corrections)

	// Write prompt to temp file to avoid shell escaping issues
	tmpPromptFile := filepath.Join(os.TempDir(), fmt.Sprintf("forge-prompt-%s.txt", util.RandomHex(4)))
	if err := util.WriteText(tmpPromptFile, prompt); err != nil {
		return nil, fmt.Errorf("failed to write prompt file: %w", err)
	}
	defer os.Remove(tmpPromptFile)

	// Pipe prompt to claude -p --output-format json
	catCmd := exec.Command("cat", tmpPromptFile)
	claudeCmd := exec.Command("claude", "-p", "--output-format", "json")

	pr, pw := io.Pipe()
	catCmd.Stdout = pw
	claudeCmd.Stdin = pr

	var outBuf strings.Builder
	claudeCmd.Stdout = &outBuf
	claudeCmd.Stderr = os.Stderr

	if err := catCmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start cat: %w", err)
	}
	if err := claudeCmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start claude: %w", err)
	}

	go func() {
		_ = catCmd.Wait()
		pw.Close()
	}()

	if err := claudeCmd.Wait(); err != nil {
		return nil, fmt.Errorf("claude analysis failed: %w", err)
	}

	responseText := outBuf.String()

	// claude -p --output-format json wraps output as {"type":"result","result":"..."}
	var content string
	var wrapper struct {
		Result string `json:"result"`
	}
	if err := json.Unmarshal([]byte(responseText), &wrapper); err == nil && wrapper.Result != "" {
		content = wrapper.Result
	} else {
		content = responseText
	}

	// Extract JSON from response
	jsonStr := extractJSON(content)
	if jsonStr == "" {
		truncated := content
		if len(truncated) > 500 {
			truncated = truncated[:500]
		}
		return nil, fmt.Errorf("could not find JSON in Claude response. Raw output:\n%s", truncated)
	}

	var analysis SpecAnalysis
	if err := json.Unmarshal([]byte(jsonStr), &analysis); err != nil {
		return nil, fmt.Errorf("failed to parse analysis JSON: %w", err)
	}
	analysis.PageCount = pageCount

	if analysis.ProjectName == "" || analysis.Language == "" {
		return nil, fmt.Errorf("incomplete analysis: missing project_name or language")
	}

	return &analysis, nil
}

func buildAnalysisPrompt(specPath, ext string, pageCount *int, corrections string) string {
	var readInstruction string
	if ext == ".pdf" {
		pagesToRead := 20
		if pageCount != nil {
			pagesToRead = *pageCount
		}
		if pagesToRead > 40 {
			pagesToRead = 40
		}
		readInstruction = fmt.Sprintf(`Read the PDF at "%s" (first %d pages) to understand what this project is.`, specPath, pagesToRead)
	} else {
		readInstruction = fmt.Sprintf(`Read the file at "%s" to understand what this project is.`, specPath)
	}

	correctionsClause := ""
	if corrections != "" {
		correctionsClause = fmt.Sprintf("\n\nThe user has provided these corrections to a previous analysis: \"%s\". Apply these corrections to your analysis.", corrections)
	}

	return fmt.Sprintf(`%s

Extract the following as a JSON object. Output ONLY the JSON, no explanation, no markdown formatting:

{
  "project_name": "short project name",
  "description": "1-2 sentence description of what this project does",
  "language": "typescript | javascript | python | go",
  "framework": "next | sveltekit | fastapi | django | gin | null",
  "project_type": "web-app | api | cli | library | automation | fullstack",
  "modules": ["list", "of", "main", "modules"],
  "architecture": "monolith | client-server | microservices | static-site | library",
  "sensitive_areas": "description of sensitive data or security concerns, or empty string",
  "domain_rules": "key business rules agents must follow, or empty string",
  "constraints": ["list", "of", "hard", "constraints"]
}

Infer language and framework from the spec's technology requirements. If not specified, choose the best fit based on the project type. For medical/healthcare projects, note HIPAA requirements in sensitive_areas.%s`, readInstruction, correctionsClause)
}

func extractJSON(content string) string {
	// Pattern 1: ```json ... ``` code block
	codeBlock := regexp.MustCompile("(?s)```(?:json)?\\s*(\\{.*?\\})\\s*```")
	if m := codeBlock.FindStringSubmatch(content); len(m) > 1 {
		return m[1]
	}

	// Pattern 2: raw JSON object with project_name
	rawObj := regexp.MustCompile(`(?s)\{.*?"project_name".*?\}`)
	if m := rawObj.FindString(content); m != "" {
		return m
	}

	// Pattern 3: the entire content is JSON
	var tmp any
	if err := json.Unmarshal([]byte(strings.TrimSpace(content)), &tmp); err == nil {
		return strings.TrimSpace(content)
	}

	return ""
}

// ─── Helpers ──────────────────────────────────────────────────────

func detectPageCount(pdfPath string) *int {
	// macOS: use mdls
	out, err := exec.Command("mdls", "-name", "kMDItemNumberOfPages", "-raw", pdfPath).Output()
	if err == nil {
		count, err := strconv.Atoi(strings.TrimSpace(string(out)))
		if err == nil && count > 0 {
			return &count
		}
	}

	// Fallback: pdfinfo
	out, err = exec.Command("bash", "-c", fmt.Sprintf(`pdfinfo "%s" 2>/dev/null | grep "^Pages:" | awk '{print $2}'`, pdfPath)).Output()
	if err == nil {
		count, err := strconv.Atoi(strings.TrimSpace(string(out)))
		if err == nil && count > 0 {
			return &count
		}
	}

	return nil
}

func extToFormat(ext string) string {
	switch ext {
	case ".pdf":
		return "PDF"
	case ".md", ".markdown":
		return "MD"
	default:
		return "TXT"
	}
}

func extToFormatLong(ext string) string {
	switch ext {
	case ".pdf":
		return "PDF"
	case ".md", ".markdown":
		return "Markdown"
	default:
		return "Text"
	}
}

File: src/commands/ingest.ts
import * as p from '@clack/prompts';
import chalk from 'chalk';
import { join, resolve, extname, basename } from 'node:path';
import { exists, readText, writeText, ensureDir } from '../utils/fs.js';
import { execaCommand } from 'execa';
import { copyFile, stat, readFile } from 'node:fs/promises';

interface IngestOptions {
	chunkSize?: string;
	resume?: string;
}

interface SpecAnalysis {
	project_name: string;
	description: string;
	language: string;
	framework: string | null;
	project_type: string;
	modules: string[];
	architecture: string;
	sensitive_areas: string;
	domain_rules: string;
	constraints: string[];
	page_count: number | null;
}

export async function ingest(files: string[], options: IngestOptions): Promise<void> {
	const cwd = process.cwd();

	p.intro(chalk.bold('forge') + chalk.dim(' — Spec Ingestion'));

	// Resolve and validate all spec files
	const resolvedFiles: { path: string; ext: string; name: string; sizeDisplay: string; sizeBytes: number; pageCount: number | null }[] = [];

	for (const file of files) {
		const specPath = resolve(file);
		if (!(await exists(specPath))) {
			p.cancel(`File not found: ${specPath}`);
			process.exit(1);
		}

		const ext = extname(specPath).toLowerCase();
		const supportedFormats = ['.pdf', '.md', '.txt', '.markdown'];
		if (!supportedFormats.includes(ext)) {
			p.cancel(`Unsupported format: ${ext} (${basename(specPath)}). Supported: ${supportedFormats.join(', ')}`);
			process.exit(1);
		}

		const fileStats = await stat(specPath);
		const sizeBytes = fileStats.size;
		const sizeMB = sizeBytes < 102400
			? `${(sizeBytes / 1024).toFixed(0)} KB`
			: `${(sizeBytes / (1024 * 1024)).toFixed(1)} MB`;

		let pageCount: number | null = null;
		if (ext === '.pdf') {
			pageCount = await detectPageCount(specPath);
		}

		resolvedFiles.push({ path: specPath, ext, name: basename(specPath), sizeDisplay: sizeMB, sizeBytes, pageCount });
	}

	const isMulti = resolvedFiles.length > 1;

	// Display source info
	const infoLines: string[] = [];
	if (isMulti) {
		infoLines.push(`Documents: ${chalk.cyan(String(resolvedFiles.length))}`);
		for (const f of resolvedFiles) {
			const format = f.ext === '.pdf' ? 'PDF' : f.ext === '.md' || f.ext === '.markdown' ? 'MD' : 'TXT';
			const pageInfo = f.pageCount ? ` (${f.pageCount}p)` : '';
			infoLines.push(`  ${chalk.dim('•')} ${chalk.cyan(f.name)} ${chalk.dim(`${format}, ${f.sizeDisplay}${pageInfo}`)}`);
		}
	} else {
		const f = resolvedFiles[0];
		const format = f.ext === '.pdf' ? 'PDF' : f.ext === '.md' || f.ext === '.markdown' ? 'Markdown' : 'Text';
		infoLines.push(`File:     ${chalk.cyan(f.name)}`);
		infoLines.push(`Format:   ${chalk.cyan(format)}`);
		infoLines.push(`Size:     ${chalk.cyan(f.sizeDisplay)}`);
		if (f.pageCount) infoLines.push(`Pages:    ${chalk.cyan(String(f.pageCount))}`);
	}

	const chunkSize = parseInt(options.chunkSize ?? '20', 10);
	const totalPages = resolvedFiles.reduce((sum, f) => sum + (f.pageCount ?? 0), 0);
	if (totalPages > chunkSize) {
		infoLines.push(`Chunks:   ${chalk.cyan(`${Math.ceil(totalPages / chunkSize)} × ${chunkSize} pages`)}`);
	}

	p.note(infoLines.join('\n'), isMulti ? 'Source Documents' : 'Source Document');

	// Generate spec ID
	const specId = `spec-${randomHex(4)}`;

	// Create spec directory and copy all files
	const specDir = join(cwd, '.forge', 'specs', specId);
	await ensureDir(specDir);

	const sourceFiles: string[] = [];
	for (let i = 0; i < resolvedFiles.length; i++) {
		const f = resolvedFiles[i];
		// Prefix with index to preserve order for multi-file
		const destName = isMulti ? `source-${i + 1}${f.ext}` : `source${f.ext}`;
		await copyFile(f.path, join(specDir, destName));
		sourceFiles.push(destName);
	}

	// If multiple text/markdown files, create a concatenated version for analysis
	let combinedPath: string | null = null;
	if (isMulti) {
		const parts: string[] = [];
		for (const f of resolvedFiles) {
			if (f.ext === '.pdf') {
				parts.push(`\n\n---\n# [PDF Document: ${f.name}]\n# (Read separately via PDF reader)\n---\n\n`);
			} else {
				const content = await readFile(f.path, 'utf-8');
				parts.push(`\n\n---\n# Document: ${f.name}\n---\n\n${content}`);
			}
		}
		combinedPath = join(specDir, 'combined.md');
		await writeText(combinedPath, parts.join(''));
	}

	p.log.success(`Copied ${resolvedFiles.length} file${isMulti ? 's' : ''} to ${chalk.dim(`.forge/specs/${specId}/`)}`);
	if (combinedPath) {
		p.log.success(`Combined document created at ${chalk.dim(`.forge/specs/${specId}/combined.md`)}`);
	}

	// Write spec metadata
	const meta = {
		spec_id: specId,
		source: {
			files: resolvedFiles.map((f, i) => ({
				original: f.name,
				stored: sourceFiles[i],
				format: f.ext.replace('.', ''),
				size_mb: parseFloat((f.sizeBytes / (1024 * 1024)).toFixed(2)),
				pages: f.pageCount,
			})),
			combined: combinedPath ? 'combined.md' : null,
			ingested_at: new Date().toISOString(),
		},
		status: 'pending-analysis',
		chunk_size: chunkSize,
	};
	await writeText(join(specDir, 'meta.json'), JSON.stringify(meta, null, 2));

	// Check if harness exists
	const harnessExists = await exists(join(cwd, 'forge.yaml'));

	if (!harnessExists) {
		const spinner = p.spinner();
		spinner.start('Analyzing spec with Claude Code...');

		let analysis: SpecAnalysis | null = null;
		// Use copied files in .forge/specs/ (not originals) so Claude can always access them
		const analysisTarget = combinedPath ?? join(specDir, sourceFiles[0]);
		const analysisExt = combinedPath ? '.md' : resolvedFiles[0].ext;
		const analysisPagesCount = combinedPath ? null : resolvedFiles[0].pageCount;

		try {
			analysis = await analyzeSpecWithClaude(analysisTarget, analysisExt, analysisPagesCount);
			spinner.stop('Spec analysis complete');
		} catch (err) {
			spinner.stop('Spec analysis failed');
			p.log.warn(chalk.yellow('Could not analyze spec automatically. You can configure manually.'));
			p.log.warn(chalk.dim(String(err)));
		}

		if (analysis) {
			const extractedLines = [
				`Project:      ${chalk.cyan(analysis.project_name)}`,
				`Description:  ${chalk.cyan(analysis.description)}`,
				`Type:         ${chalk.cyan(analysis.project_type)}`,
				`Language:     ${chalk.cyan(analysis.language)}`,
			];
			if (analysis.framework) extractedLines.push(`Framework:    ${chalk.cyan(analysis.framework)}`);
			if (analysis.modules.length > 0) extractedLines.push(`Modules:      ${chalk.cyan(analysis.modules.join(', '))}`);
			extractedLines.push(`Architecture: ${chalk.cyan(analysis.architecture)}`);
			if (analysis.sensitive_areas) extractedLines.push(`Sensitive:    ${chalk.cyan(analysis.sensitive_areas)}`);
			if (analysis.constraints.length > 0) extractedLines.push(`Constraints:  ${chalk.cyan(analysis.constraints.slice(0, 3).join('; '))}`);

			p.note(extractedLines.join('\n'), 'Extracted from spec');

			const confirmed = await p.confirm({ message: 'Does this look right?', initialValue: true });
			if (p.isCancel(confirmed)) { p.cancel('Cancelled.'); process.exit(0); }

			if (!confirmed) {
				const correctionsAnswer = await p.text({
					message: 'What needs to change?',
					placeholder: 'e.g. Use SvelteKit instead of Next.js, add billing as a module',
				});
				if (p.isCancel(correctionsAnswer)) { p.cancel('Cancelled.'); process.exit(0); }

				if (correctionsAnswer) {
					const spinner2 = p.spinner();
					spinner2.start('Re-analyzing with corrections...');
					try {
						analysis = await analyzeSpecWithClaude(analysisTarget, analysisExt, analysisPagesCount, correctionsAnswer as string);
						spinner2.stop('Updated analysis complete');
					} catch {
						spinner2.stop('Re-analysis failed, using original analysis');
					}
				}
			}

			await writeText(join(specDir, 'analysis.json'), JSON.stringify(analysis, null, 2));
			p.log.success('Spec analysis saved.');
		}
	}

	// Show next steps
	if (harnessExists) {
		p.log.step('Next step:');
		p.log.message(`  Open Claude Code → ${chalk.cyan(`/ingest ${specId}`)}`);
	} else {
		p.log.step('Next steps:');
		p.log.message(`  1. Run ${chalk.cyan('forge init')} to scaffold the harness`);
		p.log.message(`     ${chalk.dim('(use the spec analysis above to answer onboarding questions)')}`);
		p.log.message(`  2. Open Claude Code → ${chalk.cyan(`/ingest ${specId}`)}`);
	}

	p.outro(chalk.green('Spec ready for analysis.'));
}

/**
 * Called from `forge init --spec <file>` to analyze a spec and return
 * structured metadata that pre-fills the onboarding answers.
 */
export async function analyzeSpecForInit(specPath: string): Promise<SpecAnalysis | null> {
	const ext = extname(specPath).toLowerCase();
	let pageCount: number | null = null;
	if (ext === '.pdf') {
		pageCount = await detectPageCount(specPath);
	}
	return analyzeSpecWithClaude(specPath, ext, pageCount);
}

async function analyzeSpecWithClaude(
	specPath: string,
	ext: string,
	pageCount: number | null,
	corrections?: string,
): Promise<SpecAnalysis> {
	const prompt = buildAnalysisPrompt(specPath, ext, pageCount, corrections);

	// Write prompt to a temp file to avoid shell escaping issues
	const tmpPromptFile = join('/tmp', `forge-prompt-${randomHex(4)}.txt`);
	await writeText(tmpPromptFile, prompt);

	try {
		// Call claude as a subprocess, piping the prompt via stdin
		const result = await execaCommand(
			`cat "${tmpPromptFile}" | claude -p --output-format json`,
			{ timeout: 180000, shell: true },
		);

		const responseText = result.stdout;

		// claude -p --output-format json wraps output as:
		// {"type":"result","result":"<the actual content>", ...}
		let content: string;
		try {
			const wrapper = JSON.parse(responseText);
			content = wrapper.result ?? responseText;
		} catch {
			// If it's not valid JSON wrapper, use raw output
			content = responseText;
		}

		// Extract the JSON object from Claude's response
		// Try multiple patterns since Claude might wrap it in markdown code blocks
		let jsonStr: string | null = null;

		// Pattern 1: ```json ... ``` code block
		const codeBlockMatch = content.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
		if (codeBlockMatch) {
			jsonStr = codeBlockMatch[1];
		}

		// Pattern 2: raw JSON object with project_name
		if (!jsonStr) {
			const rawMatch = content.match(/\{[\s\S]*?"project_name"[\s\S]*?\}/);
			if (rawMatch) {
				jsonStr = rawMatch[0];
			}
		}

		// Pattern 3: the entire content is JSON
		if (!jsonStr) {
			try {
				JSON.parse(content);
				jsonStr = content;
			} catch { /* not json */ }
		}

		if (!jsonStr) {
			throw new Error(`Could not find JSON in Claude response. Raw output:\n${content.slice(0, 500)}`);
		}

		const analysis: SpecAnalysis = JSON.parse(jsonStr);
		analysis.page_count = pageCount;

		// Validate required fields
		if (!analysis.project_name || !analysis.language) {
			throw new Error(`Incomplete analysis: missing project_name or language`);
		}

		return analysis;
	} finally {
		// Clean up temp file
		try {
			const { unlink } = await import('node:fs/promises');
			await unlink(tmpPromptFile);
		} catch { /* ignore */ }
	}
}

function buildAnalysisPrompt(
	specPath: string,
	ext: string,
	pageCount: number | null,
	corrections?: string,
): string {
	let readInstruction: string;
	if (ext === '.pdf') {
		const pagesToRead = pageCount && pageCount > 40 ? 40 : pageCount ?? 20;
		readInstruction = `Read the PDF at "${specPath}" (first ${pagesToRead} pages) to understand what this project is.`;
	} else {
		readInstruction = `Read the file at "${specPath}" to understand what this project is.`;
	}

	let correctionsClause = '';
	if (corrections) {
		correctionsClause = `\n\nThe user has provided these corrections to a previous analysis: "${corrections}". Apply these corrections to your analysis.`;
	}

	return `${readInstruction}

Extract the following as a JSON object. Output ONLY the JSON, no explanation, no markdown formatting:

{
  "project_name": "short project name",
  "description": "1-2 sentence description of what this project does",
  "language": "typescript | javascript | python | go",
  "framework": "next | sveltekit | fastapi | django | gin | null",
  "project_type": "web-app | api | cli | library | automation | fullstack",
  "modules": ["list", "of", "main", "modules"],
  "architecture": "monolith | client-server | microservices | static-site | library",
  "sensitive_areas": "description of sensitive data or security concerns, or empty string",
  "domain_rules": "key business rules agents must follow, or empty string",
  "constraints": ["list", "of", "hard", "constraints"]
}

Infer language and framework from the spec's technology requirements. If not specified, choose the best fit based on the project type. For medical/healthcare projects, note HIPAA requirements in sensitive_areas.${correctionsClause}`;
}

async function detectPageCount(pdfPath: string): Promise<number | null> {
	try {
		const result = await execaCommand(`mdls -name kMDItemNumberOfPages -raw "${pdfPath}"`, { shell: true });
		const count = parseInt(result.stdout.trim(), 10);
		if (!isNaN(count) && count > 0) return count;
	} catch { /* fall through */ }

	try {
		const result = await execaCommand(`pdfinfo "${pdfPath}" 2>/dev/null | grep "^Pages:" | awk '{print $2}'`, { shell: true });
		const count = parseInt(result.stdout.trim(), 10);
		if (!isNaN(count) && count > 0) return count;
	} catch { /* fall through */ }

	return null;
}

function randomHex(bytes: number): string {
	const array = new Uint8Array(bytes);
	crypto.getRandomValues(array);
	return Array.from(array, (b) => b.toString(16).padStart(2, '0')).join('');
}
